### PR TITLE
Added data to harvest

### PR DESCRIPTION
**Pull request by RedVeil**

### CODE REVIEW

File: src/strategies/BaseStrategy.sol [hunk 1]
```diff
@@ -217,22 +215,44 @@ abstract contract BaseStrategy is
...
     }
 
-    function harvest() public virtual {}
+    function harvest(bytes memory data) external virtual onlyKeeperOrOwner {}
+
+    function pushFunds(
+        uint256 assets,
+        bytes memory data
+    ) external virtual onlyKeeperOrOwner {
+        _protocolDeposit(assets, convertToShares(assets), data);
```
Comment (0xruhum) [inline src/strategies/BaseStrategy.sol:239]: is it really necessary to convert assets to shares here? The inheriting contracts can do it themselves if necessary. Waste of gas imo.

File: src/strategies/BaseStrategy.sol [hunk 2]
```diff
@@ -33,18 +33,18 @@ abstract contract BaseStrategy is
...
      * @param asset_ The underlying asset used for deposit/withdraw and accounting
      * @param owner_ Owner of the contract. Controls management functions.
-     * @param autoHarvest_ Controls if the harvest function gets called on deposit/withdrawal
+     * @param autoDeposit_ Controls if `protocolDeposit` gets called on deposit
      */
     function __BaseStrategy_init(
         address asset_,
         address owner_,
-        bool autoHarvest_
+        bool autoDeposit_
```
Comment (0xruhum) [inline src/strategies/BaseStrategy.sol:41]: does this affect any of our deployment scripts?

File: src/strategies/aave/aaveV3/AaveV3Depositor.sol [hunk 3]
```diff
@@ -124,7 +128,7 @@ contract AaveV3Depositor is BaseStrategy {
     //////////////////////////////////////////////////////////////*/
 
     /// @notice Claim additional rewards given that it's active.
-    function claim() public override returns (bool success) {
+    function claim() internal override returns (bool success) {
```
Comment (0xruhum) [inline src/strategies/aave/aaveV3/AaveV3Depositor.sol:131]: are we sure we don't call `claim()` from any external contract?

File: src/strategies/curve/gauge/mainnet/CurveGaugeCompounder.sol [hunk 4]
```diff
@@ -156,15 +162,17 @@ contract CurveGaugeCompounder is BaseStrategy {
...
+            amount = IERC20(asset()).balanceOf(address(this));
+            if (amount < minOut) revert CompoundFailed();
+
+            _protocolDeposit(amount, 0, bytes(""));
+        }
         emit Harvested();
     }
 
     address[] internal _rewardTokens;
-    uint256[] public minTradeAmounts; // ordered as in rewardsTokens()
```
Comment (0xruhum) [inline src/strategies/curve/gauge/mainnet/CurveGaugeCompounder.sol:167]: why was this removed?

**Review by 0xruhum — COMMENTED** [review]
looks fine overall

File: src/strategies/curve/gauge/mainnet/CurveGaugeCompounder.sol (hunk 4 above)
Comment (RedVeil) [reply src/strategies/curve/gauge/mainnet/CurveGaugeCompounder.sol:167]: Since `harvest` is now a function only callable by authorised users the assumption was that they will take care of calling it only if its useful. The default scenario would be a keeper bot running a script that checks minTrade amounts offchain

File: src/strategies/aave/aaveV3/AaveV3Depositor.sol (hunk 3 above)
Comment (RedVeil) [reply src/strategies/aave/aaveV3/AaveV3Depositor.sol:131]: yup. only gets called in `harvest`

File: src/strategies/BaseStrategy.sol (hunk 2 above)
Comment (RedVeil) [reply src/strategies/BaseStrategy.sol:41]: it certainly does. ill adjust them once i merge all the changes in

File: src/strategies/BaseStrategy.sol (hunk 1 above)
Comment (RedVeil) [reply src/strategies/BaseStrategy.sol:239]: `_protocolDeposit` might use assets or shares depending on the implementation so i wanted to provide both so the specific implementation doesnt need to know about this function.
